### ISSUE TITLE
http pipelining is not ok

### ISSUE THREAD

**Issue by sun5858** [OISF/libhtp#304]
when pipelining is using, suricata cant output json rightly. and what is the time?

**Comment by catenacyber** [comment]
What do you mean ?
HTTP 1.1 request pipelining is supposed to be supported.

Do you a pcap sample that shows otherwise ?

**Comment by sun5858** [comment]
yes, I have
[1.pcap.zip](https://github.com/OISF/libhtp/files/5194277/1.pcap.zip)

**Comment by catenacyber** [comment]
Thanks for this pcap.

Running latest master on this I seem to get correct logs

My commands are
`./src/suricata -r 1.pcap -l log` and
`jq 'select(.event_type=="http")|.http' log/eve.json `

And the log seems correct, what did you expect ?
```
{
  "hostname": "120.52.186.72",
  "url": "/view/index.shtml",
  "http_user_agent": "Mozilla/5.0 (compatible; Nmap Scripting Engine; https://nmap.org/book/nse.html)",
  "http_content_type": "text/html",
  "http_method": "HEAD",
  "protocol": "HTTP/1.1",
  "status": 200,
  "length": 2927
}
{
  "hostname": "120.52.186.72",
  "url": "/phpmyadmin/",
  "http_user_agent": "Mozilla/5.0 (compatible; Nmap Scripting Engine; https://nmap.org/book/nse.html)",
  "http_content_type": "text/html",
  "http_method": "HEAD",
  "protocol": "HTTP/1.1",
  "status": 200,
  "length": 0
}
```

**Comment by sun5858** [comment]
Thanks for your reply so much. In pipelining mode, client does "request1->request2->request3...", and server does "response1->response2->response3...",if we want suricata detects request1 when response1 arrives, detects request2 when response2 arrives and so on in "inline=false" mode, what should I do?

**Comment by catenacyber** [comment]
I am not sure I understand your question.

> if we want suricata detects request1 when response1 arrives, detects request2 when response2 arrives and so on in "inline=false" mode, what should I do?

By default, inline=false, and Suricata does detect requests and responses and matches them together according to the logs quoted above. So, there would be nothing special to do...

I suggest you try https://forum.suricata.io which may be the better place for such a question

**Comment by sun5858** [comment]
OK

**Comment by sun5858** [comment]
hey,i have a problem with request_line when run on a pcap.The context is like this previous packet is returned from IN_COPY_BYTE_OR_RETURN in htp_connp_REQ_FINALIZE and next packet is in state of htp_connp_REQ_FINALIZE when coming, then connp->in_current_read_offset is setted to zero as connp->in_current_read_offset<len, so some content is in duplicate. The pcap is below here.
[a.pcap.zip](https://github.com/OISF/libhtp/files/5491723/a.pcap.zip)

**Comment by catenacyber** [comment]
I do not understand the problem.
Could you tell me what you would expect from Suricata logs

After running `suricata -r a.pcap -l log`
it seems to me that the logs look normal by running
`jq 'select(.event_type=="http") | .http.url, .http.status' log/eve.json`
I get
```
"/postinfo.html"
404
"/_vti_bin/_vti_aut/author.dll"
404
...
```

Looks quite similar to Wireshark analysis cf `tshark -Tfields -e  http.request.uri -e http.response.code -r a.pcap`

**Comment by sun5858** [comment]
![屏幕快照 2020-11-05 下午5 53 54](https://user-images.githubusercontent.com/52417056/98225717-097ef900-1f90-11eb-8784-833b5438ab2f.png)

jq 'select(.event_type=="http") | .http.url, .http.protocol, .http.status' log/eve.json
I get

"/postinfo.html"
"HTTP/1.1"
404
"/_vti_bin/_vti_aut/author.dll"
"HTTP/1.1"
404
"/_vti_bin/_vti_aut/author.exe"
"HTTP/1.1"
404
"/_vti_bin/_vti_aut/dvwssr.dll"
"HTTP/1.1"
404
"/_vti_bin/_vti_adm/admin.dll"
"HTTP/1.1"
404
"/_vti_bin/_vti_adm/admin.exe"
"HTTP/1.1"
404
"/_vti_bin/fpcount.exe?Page=default.asp|Image=3"
"HTTP/1.1"
404
"/_vti_bin/shtml.dll"
"HTTP/1.1"
404
"/_vti_bin/shtml.exe"
"HTTP/1.1"
404
"/_vti_pvt/_x_todo.htm"
"HTTP/1.1"
404
"/_vti_pvt/_x_todoh.htm"
"HTTP/1.1"
404
"/_vti_pvt/access.cnf"
"HTTP/1.1"
404
"/_vti_pvt/administrator.pwd"
"HTTP/1.1"
404
"/_vti_pvt/administrators.pwd"
"HTTP/1.1"
404
"/_vti_pvt/authors.pwd"
"HTTP/1.1"
404
"/_vti_pvt/bots.cnf"
"HTTP/1.1"
404
"/_vti_pvt/botinfs.cnf"
"HTTP/1.1\r\nT /_vti_pvt/botinfs.cnf HTTP/1.1"
404

**Comment by catenacyber** [comment]
Thanks, reproducing the bug with `jq 'select(.event_type=="http") | .http.protocol' log/eve.json` indeed

**Comment by sun5858** [comment]
And I have fixed it, can you help me make sure it is right or not? this is the code segment in htp_connp_REQ_FINALIZE,thanks
![屏幕快照 2020-11-05 下午6 56 10](https://user-images.githubusercontent.com/52417056/98232615-d3924280-1f98-11eb-99c3-30da0abcf49f.png)
![屏幕快照 2020-11-05 下午6 56 46](https://user-images.githubusercontent.com/52417056/98232631-d7be6000-1f98-11eb-91ea-87d426434edf.png)

**Comment by catenacyber** [comment]
I proposed another fix in #312

**Comment by victorjulien** [comment]
Closed by #312
